### ISSUE TITLE
JVM Zarr implementation?

### ISSUE THREAD

**Issue by ryan-williams** [zarr-developers/community#15]
There isn't one, is there?

I've started making one, will post updates here.

**Comment by joshmoore** [comment]
Just ran across https://github.com/bcdev/jzarr/blob/master/docs/tutorial.rst

cc: @SabineEmbacher

**Comment by SabineEmbacher** [comment]
see https://jzarr.readthedocs.io/en/latest/

hugs
Sabine


--------------------------------------------------------------------------------
Sabine Embacher
Brockmann Consult GmbH
phone: +49 (0)40 69 63 89 - 330
email: sabine.embacher@brockmann-consult.de
skype-id: sabine.embacher.bc
--------------------------------------------------------------------------------
Brockmann Consult GmbH
Chrysanderstr. 1
D-21029 Hamburg, Germany
Amtsgericht Hamburg HRB 157689
Geschäftsführer Dr. Carsten Brockmann
Web: www.brockmann-consult.de
Twitter: @BrockmannCon
--------------------------------------------------------------------------------


Am 30.03.2020 um 14:37 schrieb Josh Moore:

**Comment by SabineEmbacher** [comment]
If you need array objects which behave almost like NumPy arrays you also can wrap the data using ND4J INDArray from deeplearning4j.org. You can find examples in the data writing and reading examples.

https://jzarr.readthedocs.io/en/latest/tutorial.html#writing-and-reading-data

Or directly in the code example
https://github.com/bcdev/jzarr/blob/master/docs/examples/java/Tutorial_rtd.java#L41

**Comment by SabineEmbacher** [comment]
Can any of you tell me how to register the jzarr java library to the maven central repository. I've never done this before.
Does any of you have the time to guide or support me?

Best Regards
Sabine

**Comment by joshmoore** [comment]
Hi @SabineEmbacher. I don't remember what HOWTO we followed originally for our jars (cc: @sbesson) but https://stackoverflow.com/questions/28846802/how-to-manually-publish-jar-to-maven-central looks reasonable enough. The biggest hurdles I remember are (1) proving that you own your groupId (`*.bc.com`) and (2) making sure that all of your dependencies are accessible from maven central.  I've created https://github.com/bcdev/jzarr/issues/4 since this may become protracted, but certainly happy to help. ~Josh

**Comment by sbesson** [comment]
Following-up on https://github.com/zarr-developers/community/issues/15#issuecomment-610345738, the process used by OME for releasing some of its Java components to Sonatype is documented [here](https://docs.openmicroscopy.org/contributing/java-development.html#release-process) with the relevant links to OSSRH in case it's useful. If possible, big :+1: for having `jzarr` available from Maven Central.

**Comment by SabineEmbacher** [comment]
alimanfoo commented on 1 Aug 2018
> JVM implementation of Zarr would be very cool, particularly if it had the same flexibility as the Python implementation to plug in different storage back-ends including cloud object stores.

Did you see the example of how to read and write to Amazon AWS S3 cloud storage using JZarr?
See:
https://jzarr.readthedocs.io/en/latest/amazonS3.html
and code example
https://github.com/bcdev/jzarr/blob/master/docs/examples/java/S3Array_nio.java

**Comment by axtimwalde** [comment]
Completely missed this thread but wanted to mention that https://github.com/saalfeldlab/n5-zarr implements https://zarr.readthedocs.io/en/stable/spec/v2.html as an N5 backend since September 2019. This way it is available for array processing with ImgLib2 https://github.com/saalfeldlab/n5-imglib2 which has no size limits and built in memory caching, and is also the native data library for BigDataViewer and a bunch of processing tools that we use and build. n5-zarr includes blosc compression and locking and is included in the standard distribution of https://fiji.sc/. With the N5-API, talking to Zarr, N5, HDF5 is all the same.

There is currently no official cloud backend (other than through FS wrappers) for N5-Zarr because we haven't yet separated the interfaces for store and translation layers, i.e. writing a backend for HDF5 or Zarr is entangled with writing a backend for another store (like the AWS and GoogleCloud stores for N5).  I remember that there was a fork that copied the n5-aws-s3 logic into n5-zarr as a temporary solution @joshmoore wasn't that you who did this?

**Comment by bogovicj** [comment]
>  I remember that there was a fork that copied the n5-aws-s3 logic into n5-zarr as a temporary solution @joshmoore wasn't that you who did this?

Yup, see https://github.com/saalfeldlab/n5-aws-s3/issues/10 and https://github.com/saalfeldlab/n5-zarr/pull/5

**Comment by joshmoore** [comment]
Yup. It then got copied into the bdv/mobie code base for @tischi's I2K work. Having a way to unblock all of that would be great. (Note: I only copied-n-pasted the reader side of things. Writing still needs work as far as I know.)

**Comment by joshmoore** [comment]
As with the rust focus during the Feb. 10th meeting, there may be a Java-leaning to the upcoming call this Wednesday if anyone is interested in joining to chat.

cc: @SabineEmbacher @axtimwalde @DennisHeimbigner @WardF

**Comment by axtimwalde** [comment]
Thanks @joshmoore! I'll be there. Looking forward to seeing you all.
